### PR TITLE
Optimize(eos_designs): Set topology peer facts

### PR DESCRIPTION
**Pull request by ClausHolbechArista**
## Change Summary

Set topology peer facts

## Component(s) name

`arista.avd.eos_designs`

### New plugin `eos_designs_set_peer_facts
This plugin runs over all hostvars in the fabric inside python. For every host it finds underlay and overlay "children" and create a dictionary with the following data model:
```yaml
avd_topology_peers:
  < spine1_hostname >: [ < l3leaf1_hostname >, < l3leaf2_hostname > ]
  < spine2_hostname >: [ < l3leaf1_hostname >, < l3leaf2_hostname > ]
  < l3leaf1_hostname >: [ < l2leaf1_hostname > ]
  < l3leaf2_hostname >: [ < l2leaf1_hostname > ]
avd_overlay_peers:
  < router_server1_hostname >: [ < l3leaf1_hostname >, < l3leaf2_hostname > ]
  < router_server2_hostname >: [ < l3leaf1_hostname >, < l3leaf2_hostname > ]
```

## How to test

No changes to molecule.

#### This PR
```
arista.avd.eos_designs : Generate device configuration in structured format ------------- 260.83s
arista.avd.eos_designs : Generate fabric documentation ---------------------------------- 35.54s
arista.avd.eos_designs : Generate fabric topology in csv format. ------------------------ 29.34s
arista.avd.eos_designs : Set AVD facts -------------------------------------------------- 17.70s
arista.avd.eos_designs : Generate fabric point-to-point links summary in csv format. ---- 16.27s
arista.avd.eos_designs : Set AVD topology facts ----------------------------------------- 13.00s
arista.avd.eos_designs : Set AVD peer facts --------------------------------------------- 2.84s
arista.avd.eos_designs : Create required output directories if not present -------------- 0.43s
```
#### Devel
```
arista.avd.eos_designs : Generate device configuration in structured format ------------- 527.46s
arista.avd.eos_designs : Generate fabric documentation ---------------------------------- 35.26s
arista.avd.eos_designs : Generate fabric topology in csv format. ------------------------ 32.05s
arista.avd.eos_designs : Set AVD facts -------------------------------------------------- 18.17s
arista.avd.eos_designs : Generate fabric point-to-point links summary in csv format. ---- 17.46s
arista.avd.eos_designs : Set AVD topology facts ----------------------------------------- 12.92s
arista.avd.eos_designs : Create required output directories if not present -------------- 0.43s
```

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/avd/plugins/action/eos_designs_set_peer_facts.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+from ansible.plugins.action import ActionBase
+
+
+class ActionModule(ActionBase):
+    def run(self, tmp=None, task_vars=None):
+        if task_vars is None:
+            task_vars = {}
+
+        result = super().run(tmp, task_vars)
+        del tmp  # tmp no longer has any effect
```
Comment (xaviramon) [inline ansible_collections/arista/avd/plugins/action/eos_designs_set_peer_facts.py:13]: Really needed?

Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/action/eos_designs_set_peer_facts.py:13]: It is part of a standard layout of ansible action modules.

**Comment by xaviramon** [comment]
With this last commit I am seeing a degradation in performance of around 10% compared to previous one.
```
Wednesday 02 March 2022  15:38:30 +0100 (0:00:03.934)       0:12:54.462 ******* 
=============================================================================== 
arista.avd.eos_designs : Generate device configuration in structured format --------------------------------------------------------------------------------------- 177.19s
arista.avd.eos_designs : Set AVD facts ----------------------------------------------------------------------------------------------------------------------------- 30.48s
arista.avd.eos_designs : Set AVD topology facts -------------------------------------------------------------------------------------------------------------------- 27.89s
arista.avd.eos_designs : Generate fabric documentation ------------------------------------------------------------------------------------------------------------- 23.94s
arista.avd.eos_designs : Generate fabric point-to-point links summary in csv format. ------------------------------------------------------------------------------- 16.87s
arista.avd.eos_designs : Generate fabric topology in csv format. ---------------------------------------------------------------------------------------------------- 8.03s
arista.avd.eos_designs : Set AVD peer facts ------------------------------------------------------------------------------------------------------------------------- 4.12s


Tuesday 08 March 2022  10:36:55 +0100 (0:00:03.906)       0:14:15.783 ********* 
=============================================================================== 
arista.avd.eos_designs : Generate device configuration in structured format ---------------------------------------------------------------------------------------------- 235.72s
arista.avd.eos_designs : Set AVD facts ------------------------------------------------------------------------------------------------------------------------------------ 26.48s
arista.avd.eos_designs : Generate fabric documentation -------------------------------------------------------------------------------------------------------------------- 25.15s
arista.avd.eos_designs : Set AVD topology facts --------------------------------------------------------------------------------------------------------------------------- 21.34s
arista.avd.eos_designs : Generate fabric point-to-point links summary in csv format. -------------------------------------------------------------------------------------- 17.02s
arista.avd.eos_designs : Generate fabric topology in csv format. ----------------------------------------------------------------------------------------------------------- 8.74s
arista.avd.eos_designs : Set AVD peer facts -------------------------------------------------------------------------------------------------------------------------------- 2.82s
arista.avd.eos_designs : Create required output directories if not present ------------------------------------------------------------------------------------------------- 0.78s
```

I am still seeing an improve of 22% compared to devel.

**Review by xaviramon — APPROVED** [review]
LGTM

File: ansible_collections/arista/avd/plugins/modules/eos_designs_set_peer_facts.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+author: Arista Ansible Team (@aristanetworks)
+short_description: Set eos_designs peer facts
+description:
+  - Parse all hostvars and set a common avd_topology_peers and avd_overlay_peers dicts with list per host
+options:
+  avd_topology_peers:
+    description: |
+      Set avd_topology_peers fact with a list per device of downstream devices,
+      meaning devices pointing to this device as uplink switch
+    required: false
```
Comment (KonikaChaurasiya-GSLab) [inline ansible_collections/arista/avd/plugins/modules/eos_designs_set_peer_facts.py]: Just a minor change - shouldn't the case for true/false values be consistent either True/False or true/false

File: ansible_collections/arista/avd/roles/eos_designs/templates/inband_management/parent-logic.j2 [hunk 3]
```diff
@@ -1,15 +1,18 @@
...
-{# Add underlay links defined on other devices pointing to this device #}
-{% for fabric_switch in groups[fabric_name] | arista.avd.natural_sort %}
-{%     set fabric_switch_facts = hostvars[fabric_switch].switch | arista.avd.default %}
-{%     if inventory_hostname in fabric_switch_facts.inband_management_parents | arista.avd.default([]) %}
-{%         set inband_management_data.role = 'parent' %}
-{%         if fabric_switch_facts.inband_management_subnet not in inband_management_data.subnets %}
-{%             do inband_management_data.vlans.append(fabric_switch_facts.inband_management_vlan) %}
-{%             do inband_management_data.subnets.append(fabric_switch_facts.inband_management_subnet) %}
+{# Add inband management variables defined on other devices pointing to this device #}
+{% set avd_topology_peers = hostvars[ansible_play_hosts[0]].ansible_facts.avd_topology_peers[inventory_hostname] | arista.avd.default([]) %}
```
Comment (KonikaChaurasiya-GSLab) [inline ansible_collections/arista/avd/roles/eos_designs/templates/inband_management/parent-logic.j2:5]: Do we need to add a defined check for 'hostvars' . Like ``` if hostvars is arista.avd.defined ```

**Review by KonikaChaurasiya-GSLab — CHANGES_REQUESTED** [review]
LGTM!

**Review by manuwelakanade — APPROVED** [review]
LGTM

File: ansible_collections/arista/avd/roles/eos_designs/templates/inband_management/parent-logic.j2 (hunk 3 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/inband_management/parent-logic.j2:5]: hostvars is an "Ansible magic variable", so it is always available.

File: ansible_collections/arista/avd/plugins/modules/eos_designs_set_peer_facts.py (hunk 2 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/modules/eos_designs_set_peer_facts.py]: Fixed. Thanks

**Pull request merged**